performance-tests/bench_2/node_controller/main.cpp

```diff
@@ -548,113 +557,159 @@ int ACE_TMAIN(int argc, ACE_TCHAR* argv[])
+          if (allocated_scenario.scenario_id == TAO::String_Manager()) {
+            if (write_status(name, this_node_id, BUSY, status_writer_impl)) {
+              allocated_scenario = scenarios[scenario];
+              initial_attempt = std::chrono::system_clock::now();
             }
           }
         }
-        waiting = false;
-        break;
       }
+      if (std::string(scenarios[scenario].scenario_id.in()) == std::string(allocated_scenario.scenario_id.in()) && !(scenarios[scenario].launch_time == ZERO)) {
+        allocated_scenario.launch_time = scenarios[scenario].launch_time;
+      }
+    }
+    if (allocated_scenario.configs.length() != 0 && !(allocated_scenario.launch_time == ZERO)) {
+      complete = true;
     }
   }
+  result = allocated_scenario;
+}
 
-  // Report Busy Status
-  {
-    Status status;
-    status.node_id = this_node_id;
-    status.state = BUSY;
-    status.name = name.c_str();
-    if (status_writer_impl->write(status, DDS::HANDLE_NIL)) {
-      std::cerr << "Write status failed" << std::endl;
-      return 1;
+int run_cycle(
+  const std::string& name,
+  ACE_Process_Manager& process_manager,
+  DDS::DomainParticipant_var participant,
+  StatusDataWriter_var status_writer_impl,
+  AllocatedScenarioDataReader_var config_reader_impl,
+  ReportDataWriter_var report_writer_impl)
+{
+  NodeId this_node_id = dynamic_cast<OpenDDS::DCPS::DomainParticipantImpl*>(participant.in())->get_id();
+
+  WorkerManager worker_manager(process_manager);
+
+  // Wait for Status Publication with Test Controller and Write Status
+  if (!write_status(name, this_node_id, AVAILABLE, status_writer_impl)) {
+    std::cerr << "Write status (available) failed\n" << std::flush;
+    return 1;
+  }
+
+  Bench::TestController::AllocatedScenario scenario;
+  wait_for_full_scenario(name, this_node_id, status_writer_impl, config_reader_impl, scenario);
+
+  Bench::NodeController::Configs& configs = scenario.configs;
+  for (CORBA::ULong node = 0; node < configs.length(); ++node) {
+    if (configs[node].node_id == this_node_id) {
+      worker_manager.timeout(configs[node].timeout);
+      CORBA::ULong config_count = configs[node].workers.length();
+      for (CORBA::ULong config = 0; config < config_count; config++) {
+        WorkerId& id = configs[node].workers[config].worker_id;
+        const WorkerId end = id + configs[node].workers[config].count;
+        for (; id < end; id++) {
+          if (worker_manager.add_worker(this_node_id, configs[node].workers[config])) {
+            return 1;
+          }
+        }
+      }
+      break;
     }
   }
 
+  using Builder::ZERO;
+
+  std::chrono::time_point<std::chrono::system_clock> timeout_time;
+  if (scenario.launch_time < ZERO) {
+    timeout_time = std::chrono::system_clock::now() - get_duration(scenario.launch_time);
+  } else {
+    timeout_time = std::chrono::system_clock::time_point(get_duration(scenario.launch_time));
+  }
+  std::this_thread::sleep_until(timeout_time);
```

Comment:
Should it check if it already missed the launch time?

Reply:
since it's always calculating the absolute time, sleep_until will just return immediately for launch times in the past.

Reply:
Yea, I was wondering if that was intended, but I sounds like it is. An alternative I was thinking of was a threshold for missing the launch time before giving up. I'm guessing missing by enough to mess up the scenarios would be unlikely though.